### PR TITLE
Added error boundary to handle 404 errors

### PR DESCRIPTION
**Pull request by pricea1**
### What is the context of this PR?
Display Page Not Found page when navigating to non existant questionnaire in URL. URL is not to be changed.

### How to review 
1. Navigate to non existant questionnaire e.g. /#/q/non-existant
2. Page not found will display. nb: in Dev, the triggered JS error will display after a few seconds. The error is dismissable and only displayed in dev - this is a by design in React. see https://github.com/facebook/create-react-app/issues/3627
3. Ensure unit test passes
4. Ensure new Error_spec integration test passes.

### CODE REVIEW

**Comment by pricea1** [comment]
[Error on questionnaire doesn't exist (2)](https://trello.com/c/jBtoLGw7/999-error-on-questionnaire-doesnt-exist-2)

File: eq-author/src/App/ErrorBoundary/index.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+export default class ErrorBoundary extends React.Component {
+  static propTypes = {
+    children: PropTypes.object.isRequired, // eslint-disable-line
+  };
+
+  state = { error: false };
+
+  componentDidCatch(error, errorInfo) {
+    // Catch errors in any components below and re-render with error message
```
Comment (tgandrews) [inline eq-author/src/App/ErrorBoundary/index.js]: I don't think these comments are needed

File: eq-author/src/App/ErrorBoundary/index.test.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  beforeEach(() => {
+    wrapper = shallow(
+      <ErrorBoundary>
+        <MockComponent />
+      </ErrorBoundary>
+    );
+  });
+
+  it("should render the child component by default", () => {
+    expect(wrapper.instance().render()).toMatchObject(<MockComponent />);
```
Comment (tgandrews) [inline eq-author/src/App/ErrorBoundary/index.test.js]: Could this be something like this?
`expect(wrapper.find(MockComponent)).toHaveLength(1)`

File: eq-author/src/App/QuestionnaireDesignPage/index.js [hunk 3]
```diff
@@ -215,8 +217,15 @@ export default withMutations(props => (
       },
     }}
   >
-    {innerProps => (
-      <UnwrappedQuestionnaireDesignPage {...innerProps} {...props} />
-    )}
+    {innerProps => {
+      if (!innerProps.data || !innerProps.data.questionnaire) {
```
Comment (tgandrews) [inline eq-author/src/App/QuestionnaireDesignPage/index.js]: There are some code paths that check for `loading` in `render` that I don't think will ever run now. As we will never have data whilst we are loading.

I presume we still want to show loading messages

File: eq-author/src/App/ErrorBoundary/index.js (hunk 1 above)
Comment (pricea1) [reply eq-author/src/App/ErrorBoundary/index.js]: now removed

File: eq-author/src/App/ErrorBoundary/index.test.js (hunk 2 above)
Comment (pricea1) [reply eq-author/src/App/ErrorBoundary/index.test.js]: yes. code updated

File: eq-author/src/App/QuestionnaireDesignPage/index.js (hunk 3 above)
Comment (pricea1) [reply eq-author/src/App/QuestionnaireDesignPage/index.js]: refactored so <UnwrappedQuestionnaireDesignPage /> always returned unless component not loading when no questionnaire data.

File: eq-author/cypress/integration/authenticated/errors_spec.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  });
+
+  it("displays page not found page if navigating to non-existant questionnaire", done => {
+    cy.on("uncaught:exception", () => {
+      // return false to prevent the error from
+      // failing this test
+      return false;
+    });
+
+    cy.visit("http://localhost:3000/#/q/not-existant-id");
```
Comment (tgandrews) [inline eq-author/cypress/integration/authenticated/errors_spec.js]: Ah this domain driven by an environment variable which is defined here https://github.com/ONSdigital/eq-author-app/blob/e85f6f9d2ef9b9e86e59bcebee4722aabeffd7f2/eq-author/.env.test#L1

However you would probably want to read it with https://docs.cypress.io/guides/references/configuration.html#Cypress-config

File: eq-author/src/App/ErrorBoundary/index.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import React from "react";
+import PropTypes from "prop-types";
+import NotFound from "App/NotFoundPage";
+import { ERR_PAGE_NOT_FOUND } from "constants/error-codes";
+
+export default class ErrorBoundary extends React.Component {
+  static propTypes = {
+    children: PropTypes.object.isRequired, // eslint-disable-line
```
Comment (SamGodwin2) [inline eq-author/src/App/ErrorBoundary/index.js]: Can we use `Proptypes.node` here, save having to disable the line?

File: eq-author/cypress/integration/authenticated/errors_spec.js (hunk 4 above)
Comment (pricea1) [reply eq-author/cypress/integration/authenticated/errors_spec.js]: Now changed to relative URL.

File: eq-author/src/App/ErrorBoundary/index.js (hunk 5 above)
Comment (pricea1) [reply eq-author/src/App/ErrorBoundary/index.js]: Done

File: eq-author/src/App/ErrorBoundary/index.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+  render() {
+    if (this.state.error.message === ERR_PAGE_NOT_FOUND) {
+      // You can render any custom fallback UI
+      return <NotFound />;
+    }
+
+    return this.props.children;
+  }
+}
```
Comment (samiwel) [inline eq-author/src/App/ErrorBoundary/index.js]: As discussed, just an issue here about resetting the state.

Comment (pricea1) [reply eq-author/src/App/ErrorBoundary/index.js]: Now resetting state on url pathname change and sign out

**Pull request merged**